File: frontend/types/multi-agent.ts
// Multi-Agent collaboration types

export type CollaborationMode = 'sequential' | 'parallel' | 'debate' | 'hierarchical' | 'round_robin';
export type SessionStatus = 'created' | 'running' | 'paused' | 'completed' | 'failed' | 'cancelled';
export type MemberRole = 'leader' | 'member' | 'supporter' | 'opponent' | 'judge';
export type MessageType = 'user_input' | 'agent_output' | 'system' | 'reference';

// ============== Agent Member ==============

export interface AgentMember {
  id: string;
  group_id: string;
  name: string;
  role: MemberRole;
  system_prompt: string;
  execution_order: number;
  model?: string;
  model_provider?: string;
  temperature: number;
  max_tokens: number;
  tools: any[];
  icon?: string;
  color?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface AgentMemberCreate {
  name: string;
  role: MemberRole;
  system_prompt?: string;
  execution_order?: number;
  model?: string;
  model_provider?: string;
  temperature?: number;
  max_tokens?: number;
  tools?: any[];
  icon?: string;
  color?: string;
}

export interface AgentMemberUpdate {
  name?: string;
  role?: MemberRole;
  system_prompt?: string;
  execution_order?: number;
  model?: string;
  model_provider?: string;
  temperature?: number;
  max_tokens?: number;
  tools?: any[];
  icon?: string;
  color?: string;
  is_active?: boolean;
}

// ============== Agent Group ==============

export interface ModeConfig {
  // For sequential
  stop_on_first_success?: boolean;
  
  // For parallel
  max_parallel?: number;
  aggregation_method?: 'merge' | 'vote' | 'first';
  
  // For debate
  rounds?: number;
  allow_rebuttal?: boolean;
  judge_model?: string;
  
  // For hierarchical
  max_depth?: number;
  delegate_threshold?: number;
  
  // For round_robin
  max_turns?: number;
  allow_skip?: boolean;
}

export interface AgentGroup {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  mode: CollaborationMode;
  mode_config: ModeConfig;
  default_model: string;
  default_provider: string;
  enable_orchestrator: boolean;
  orchestrator_prompt?: string;
  max_iterations: number;
  termination_prompt?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  members?: AgentMember[];
}

export interface AgentGroupCreate {
  name: string;
  description?: string;
  mode: CollaborationMode;
  mode_config?: ModeConfig;
  default_model?: string;
  default_provider?: string;
  enable_orchestrator?: boolean;
  orchestrator_prompt?: string;
  max_iterations?: number;
  termination_prompt?: string;
  members?: AgentMemberCreate[];
}

export interface AgentGroupUpdate {
  name?: string;
  description?: string;
  mode?: CollaborationMode;
  mode_config?: ModeConfig;
  default_model?: string;
  default_provider?: string;
  enable_orchestrator?: boolean;
  orchestrator_prompt?: string;
  max_iterations?: number;
  termination_prompt?: string;
  is_active?: boolean;
}

// ============== Session ==============

export interface AgentMessage {
  id: string;
  session_id: string;
  member_id?: string;
  member_name?: string;
  member_color?: string;
  message_type: MessageType;
  content: string;
  turn: number;
  iteration: number;
  referenced_message_id?: string;
  tool_calls?: any[];
  tool_results?: any[];
  model_used?: string;
  tokens_used?: number;
  execution_time_ms?: number;
  created_at: string;
}

export interface CollaborationSession {
  id: string;
  group_id: string;
  user_id: string;
  status: SessionStatus;
  initial_input: string;
  context?: Record<string, any>;
  final_output?: string;
  summary?: string;
  current_turn: number;
  completed_iterations: number;
  created_at: string;
  started_at?: string;
  completed_at?: string;
  error_message?: string;
  messages?: AgentMessage[];
}

export interface SessionStart {
  initial_input: string;
  context?: Record<string, any>;
}

export interface SessionExecuteRequest {
  user_input?: string;
  continue_session?: boolean;
}

export interface SessionExecuteResponse {
  session_id: string;
  status: SessionStatus;
  current_turn: number;
  messages: AgentMessage[];
  final_output?: string;
  is_complete: boolean;
}

// ============== Mode Descriptions ==============

export const MODE_DESCRIPTIONS: Record<CollaborationMode, { title: string; description: string; icon: string }> = {
  sequential: {
    title: 'Sequential',
    description: 'Agents execute in a fixed order, each seeing the output of previous agents',
    icon: '→'
  },
  parallel: {
    title: 'Parallel',
    description: 'Multiple agents work simultaneously on the same task, results are aggregated',
    icon: '⚡'
  },
  debate: {
    title: 'Debate',
    description: 'Agents take pro/con positions and debate, with an optional judge',
    icon: '⚖️'
  },
  hierarchical: {
    title: 'Hierarchical',
    description: 'A leader agent coordinates subordinate agents through delegation',
    icon: '📋'
  },
  round_robin: {
    title: 'Round Robin',
    description: 'Agents take turns speaking in rotation until a conclusion is reached',
    icon: '🔄'
  }
};

export const MEMBER_ROLE_COLORS: Record<MemberRole, string> = {
  leader: '#8b5cf6',    // Purple
  member: '#3b82f6',    // Blue
  supporter: '#10b981', // Green
  opponent: '#ef4444',  // Red
  judge: '#f59e0b',     // Amber
};

export const MODE_COLORS: Record<CollaborationMode, string> = {
  sequential: '#3b82f6',
  parallel: '#8b5cf6',
  debate: '#ef4444',
  hierarchical: '#10b981',
  round_robin: '#f59e0b',
};

File: backend/app/models/schemas/multi_agent.py
"""Multi-Agent collaboration schemas."""
from datetime import datetime
from typing import Optional, Any
from pydantic import BaseModel, Field


# ============== Agent Group Schemas ==============

class AgentMemberBase(BaseModel):
    """Base schema for agent member."""
    name: str = Field(..., min_length=1, max_length=100)
    role: str = Field(default="member", description="Member role: leader, member, supporter, opponent, judge")
    system_prompt: str = Field(default="You are a helpful AI assistant.")
    execution_order: int = Field(default=0, ge=0)
    model: Optional[str] = Field(None, description="Override group default model")
    model_provider: Optional[str] = Field(None, description="Override group default provider")
    temperature: float = Field(default=0.7, ge=0, le=2)
    max_tokens: int = Field(default=4096, ge=1, le=32768)
    tools: Optional[list] = Field(default_factory=list)
    icon: Optional[str] = Field(None, description="Icon name for UI")
    color: Optional[str] = Field(None, description="Color for UI visualization")


class AgentMemberCreate(AgentMemberBase):
    """Schema for creating an agent member."""
    pass


class AgentMemberUpdate(BaseModel):
    """Schema for updating an agent member."""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    role: Optional[str] = None
    system_prompt: Optional[str] = None
    execution_order: Optional[int] = Field(None, ge=0)
    model: Optional[str] = None
    model_provider: Optional[str] = None
    temperature: Optional[float] = Field(None, ge=0, le=2)
    max_tokens: Optional[int] = Field(None, ge=1, le=32768)
    tools: Optional[list] = None
    icon: Optional[str] = None
    color: Optional[str] = None
    is_active: Optional[bool] = None


class AgentMemberResponse(AgentMemberBase):
    """Schema for agent member response."""
    id: str
    group_id: str
    is_active: bool
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class AgentGroupBase(BaseModel):
    """Base schema for agent group."""
    name: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    mode: str = Field(default="sequential", description="Collaboration mode")
    mode_config: Optional[dict] = Field(default_factory=dict)
    default_model: str = Field(default="llama2")
    default_provider: str = Field(default="ollama")
    enable_orchestrator: bool = Field(default=True)
    orchestrator_prompt: Optional[str] = None
    max_iterations: int = Field(default=10, ge=1, le=100)
    termination_prompt: Optional[str] = None


class AgentGroupCreate(AgentGroupBase):
    """Schema for creating an agent group."""
    members: Optional[list[AgentMemberCreate]] = Field(default_factory=list)


class AgentGroupUpdate(BaseModel):
    """Schema for updating an agent group."""
    name: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = Field(None, max_length=1000)
    mode: Optional[str] = None
    mode_config: Optional[dict] = None
    default_model: Optional[str] = None
    default_provider: Optional[str] = None
    enable_orchestrator: Optional[bool] = None
    orchestrator_prompt: Optional[str] = None
    max_iterations: Optional[int] = Field(None, ge=1, le=100)
    termination_prompt: Optional[str] = None
    is_active: Optional[bool] = None


class AgentGroupResponse(AgentGroupBase):
    """Schema for agent group response."""
    id: str
    user_id: str
    is_active: bool
    created_at: datetime
    updated_at: datetime
    members: Optional[list[AgentMemberResponse]] = None
    
    class Config:
        from_attributes = True


class AgentGroupListResponse(BaseModel):
    """Schema for paginated agent group list response."""
    items: list[AgentGroupResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


# ============== Session Schemas ==============

class SessionStart(BaseModel):
    """Schema for starting a collaboration session."""
    initial_input: str = Field(..., min_length=1)
    context: Optional[dict] = Field(default_factory=dict)


class SessionUpdate(BaseModel):
    """Schema for updating a session."""
    status: Optional[str] = None
    context: Optional[dict] = None
    final_output: Optional[str] = None
    summary: Optional[str] = None


class AgentMessageBase(BaseModel):
    """Base schema for agent message."""
    message_type: str = Field(default="agent_output")
    content: str
    referenced_message_id: Optional[str] = None
    tool_calls: Optional[list] = None
    tool_results: Optional[list] = None
    model_used: Optional[str] = None
    tokens_used: Optional[int] = None
    execution_time_ms: Optional[int] = None
    turn: int = Field(default=0)
    iteration: int = Field(default=0)


class AgentMessageCreate(AgentMessageBase):
    """Schema for creating an agent message."""
    member_id: Optional[str] = None


class AgentMessageResponse(AgentMessageBase):
    """Schema for agent message response."""
    id: str
    session_id: str
    member_id: Optional[str] = None
    member_name: Optional[str] = None
    member_color: Optional[str] = None
    created_at: datetime
    
    class Config:
        from_attributes = True


class SessionResponse(BaseModel):
    """Schema for collaboration session response."""
    id: str
    group_id: str
    user_id: str
    status: str
    initial_input: str
    context: Optional[dict] = None
    final_output: Optional[str] = None
    summary: Optional[str] = None
    current_turn: int
    completed_iterations: int
    created_at: datetime
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    error_message: Optional[str] = None
    messages: Optional[list[AgentMessageResponse]] = None
    
    class Config:
        from_attributes = True


class SessionListResponse(BaseModel):
    """Schema for paginated session list response."""
    items: list[SessionResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


# ============== Execution Schemas ==============

class SessionExecuteRequest(BaseModel):
    """Schema for executing a collaboration session."""
    user_input: Optional[str] = Field(None, description="Additional user input")
    continue_session: bool = Field(default=True, description="Continue existing session or create new")


class SessionExecuteResponse(BaseModel):
    """Schema for session execution response."""
    session_id: str
    status: str
    current_turn: int
    messages: list[AgentMessageResponse]
    final_output: Optional[str] = None
    is_complete: bool = False


# ============== Mode Configuration Schemas ==============

class SequentialModeConfig(BaseModel):
    """Configuration for sequential mode."""
    stop_on_first_success: bool = Field(default=False)


class ParallelModeConfig(BaseModel):
    """Configuration for parallel mode."""
    max_parallel: int = Field(default=3)
    aggregation_method: str = Field(default="merge", description="merge, vote, or first")


class DebateModeConfig(BaseModel):
    """Configuration for debate mode."""
    rounds: int = Field(default=3, ge=1, le=10)
    allow_rebuttal: bool = Field(default=True)
    judge_model: Optional[str] = Field(None, description="Model for judge agent")


class HierarchicalModeConfig(BaseModel):
    """Configuration for hierarchical mode."""
    max_depth: int = Field(default=3, ge=1, le=5)
    delegate_threshold: float = Field(default=0.5, description="Confidence threshold to delegate")


class RoundRobinModeConfig(BaseModel):
    """Configuration for round robin mode."""
    max_turns: int = Field(default=10, ge=1, le=50)
    allow_skip: bool = Field(default=True)
